velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,286 @@
+    }
+    bool fixedMinFrame = minFrame == frameStarts[0];
+
+    vector_size_t numFrameRows = maxFrame + 1 - minFrame;
+    for (int i = 0; i < argIndices_.size(); i++) {
+      argVectors_[i]->resize(numFrameRows);
+      partition_->extractColumn(
+          argIndices_[i], minFrame, numFrameRows, 0, argVectors_[i]);
+    }
+
+    return {minFrame, maxFrame, fixedMinFrame, increasingFrame};
+  }
+
+  void fixedStartAggregation(
+      vector_size_t numRows,
+      vector_size_t minFrame,
+      vector_size_t maxFrame,
+      const vector_size_t* frameEndsVector,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    SelectivityVector rows;
+    rows.resize(maxFrame + 1 - minFrame);
+
+    auto singleGroup = std::vector<vector_size_t>{0};
+    // Initialize the aggregate at the beginning of the output block.
+    aggregate_->clear();
+    aggregate_->initializeNewGroups(&singleGroupRow_, singleGroup);
+    auto frameBeginIndex = minFrame;
+
+    // This is a simple optimization for blocks that have a fixed minFrame
+    // and increasing frameEnd values. In that case, we can
+    // incrementally aggregate over the new rows seen in the frame between
+    // the previous and current row.
+    for (int i = 0; i < numRows; i++) {
+      auto frameEndIndex = frameEndsVector[i] - minFrame + 1;
+      if (frameEndIndex > frameBeginIndex) {
+        rows.clearAll();
+        rows.setValidRange(frameBeginIndex, frameEndIndex, true);
+        rows.updateBounds();
+
+        aggregate_->addSingleGroupRawInput(
+            singleGroupRow_, rows, argVectors_, false);
+        aggregate_->extractValues(&singleGroupRow_, 1, &aggregateResultVector_);
+      }
+
+      result->copy(aggregateResultVector_.get(), resultOffset + i, 0, 1);
+      frameBeginIndex = frameEndIndex;
+    }
+  }
+
+  void naiveAggregation(
+      vector_size_t numRows,
+      vector_size_t minFrame,
+      vector_size_t maxFrame,
+      const vector_size_t* frameStartsVector,
+      const vector_size_t* frameEndsVector,
+      vector_size_t resultOffset,
+      const VectorPtr& result) {
+    SelectivityVector rows;
+    rows.resize(maxFrame + 1 - minFrame);
+    auto singleGroup = std::vector<vector_size_t>{0};
+    for (int i = 0; i < numRows; i++) {
+      // This is a very naive algorithm.
+      // It evaluates the entire aggregation for each row by iterating over
+      // input rows from frameStart to frameEnd in the SelectivityVector.
+      // TODO : Try to re-use previous computations by advancing and retracting
+      // the aggregation based on the frame changes with each row. This would
+      // require adding new APIs to the Aggregate framework.
+      aggregate_->clear();
+      aggregate_->initializeNewGroups(&singleGroupRow_, singleGroup);
+
+      auto frameStartIndex = frameStartsVector[i] - minFrame;
+      auto frameEndIndex = frameEndsVector[i] - minFrame + 1;
+      rows.clearAll();
```

Comment:
The code from here till aggregate_->extractValues repeats between two algorithms. Consider extracting into a helper method to avoid copy-paste.